Document next major React version

MMMalik opened the pull request:
Provides documentation for next major React version.

- Docs from v1 are moved to `v1` folder
- Docs for v2 are placed under `v2` folder
- v1 README is updated with info about new version
- v2 README is created
- redirects are added
- links referencing React v1 are updated

Closes ckeditor/ckeditor4-react#208.

MMMalik commented:
Hello @Comandeer,

there is a detailed documentation for v2 in this PR. It might help you with the review of the [main v2 PR](https://github.com/ckeditor/ckeditor4-react/pull/199).

sculpt0r reviewed (changes requested):
A huge and good piece of work :+1: 

I left a few inline comments, but also I have two thoughts:
- TBH, this react API & props are so big, that maybe they deserve to separate sections inside our API documentation. We already have a section for the jQUery adapter. It will require a lot of work with our docs repo, so we should talk about it with @f1ames ?
- I wasn't able to run `basic` and `basic-typescript` samples with `npm run build` and `npm run start`. Didn't try with others. I just wanted to check whenever I will be able to set up or change things as the documentation says. Did I forget about something?

MMMalik commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,635 @@
+
+## CKEditor component
+
+The `CKEditor` component is a solution that should satisfy most use-cases. It's a wrapper around low-level `useCKEditor` hook.
```
Agreed!

MMMalik commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,635 @@
+For licensing, see LICENSE.md.
+-->
+
+# CKEditor 4 React Integration - v2
```
Agreed!

MMMalik commented:
Thank you @sculpt0r for detailed review.

I have made the following changes as per review suggestions:

- fixed code formatting
- re-phrased highlighted paragraph
- added relevant code samples

Regarding samples:
- `react` dependencies are upgraded
- samples are refactored to functional approach
- `Two-way data binding` is renamed to `Lifting state up` because that is the correct term in React ecosystem for this kind of functionality

> TBH, this react API & props are so big, that maybe they deserve to separate sections inside our API documentation. We already have a section for the jQUery adapter. It will require a lot of work with our docs repo, so we should talk about it with @f1ames ?

That was my impression as well. Therefore, I have shortened following sections: 
- `Editor initialization` because it started to look like a generic React primer rather than docs for CKEditor React integration
- Removed docs for few functions under `Utilities` section. We don't really need to re-export anyway.

It's still ~50% longer than v1 README but please remember that I had to document `useCKEditor` hook, TypeScript support and v1 -> v2 migration.

> I wasn't able to run basic and basic-typescript samples with npm run build and npm run start. Didn't try with others. I just wanted to check whenever I will be able to set up or change things as the documentation says. Did I forget about something?

Have you consulted `Samples` section under `CONTRIBUTING` in `ckeditor4-react` repo? There is a guide on how to run a sample. Basically, you must remember to use `npm link` command. Also, please make sure that you use the freshest `major`.

sculpt0r commented:
Hi @MMMalik,

Thank you for so many corrects.

> That was my impression as well. Therefore, I have shortened following sections:
> It's still ~50% longer than v1 README but please remember that I had to document useCKEditor hook, TypeScript support and v1 -> v2 migration.

I actually don't mind that it is long. If we have a lot of things to describe - just describe them. I talked with @f1ames and we agreed that we could try export some API references to the API itself. We could start with react integration, and then do the same with the others. Some details how I see this:
- in `ckeditor4-dosc` we add submodule `git add submodule https://github.com/ckeditor/ckeditor4-react.git integrations/ckeditor4-react`
- in `ckeditorr-docs` we add the source file to parse with API `integrations/ckeditor4-react/src/doc.js`: https://github.com/ckeditor/ckeditor4-docs/blob/57a85673f251260e69798a5ec1803bc9ca9fcb0b/gruntfile.js#L155-L166

Then all API comments from react doc.js file will land at API integrations e.g. 
![image](https://user-images.githubusercontent.com/6498567/122294739-adeaf580-cef8-11eb-9fd8-066bbc57ec7f.png)
 Please take a look at how a similar section is created for jQuery adapter: https://github.com/ckeditor/ckeditor4/blob/508764d519531eb2c4df93e686d5d49235d9a9e6/adapters/jquery.js#L6-19
 
 Why this way? Because we will keep all API comments close to the original code, so inside react-integration repo. We set react-integration as a submodule to avoid using npm dependencies. In this way, we not pollute output `dist` from this library. We are using `doc.js` not comments inside `.ts` files because as far as I found - jsduck (our API comments parser) doesn't support them. So all API tables you put inside the document could be replaced with the links to the proper API section. They are linkable, there is more tooling to for example mark them as deprecated, or private. They should be easier linkable in issue comments. We have a more elegant way to display default values and descriptions. Also, we still have a place to put there some code samples. Also, you will be able to mark props and others with a proper version tag e.g. 
![image](https://user-images.githubusercontent.com/6498567/122296282-8432ce00-cefa-11eb-947f-69733dd89801.png)

There should be some proper note, that thinks under integration API aren't really classes under CKEDITOR.React namespace, but are easier to refer to in this way. 

There will be another part of this review soon ;)

sculpt0r reviewed (changes requested):
Besides: https://github.com/ckeditor/ckeditor4-docs/pull/355#issuecomment-862739954

Building docs gives us Error in https://github.com/ckeditor/ckeditor4-docs/blob/t/208/docs/guide/dev/integration/react/v2/README.md 
![image](https://user-images.githubusercontent.com/6498567/122298000-7ed68300-cefc-11eb-9291-12391ff2351d.png) Indeed - link is invalid.

I use your advice about linking to react repo, but go in browser (on samples page)
![image](https://user-images.githubusercontent.com/6498567/122337566-cc73df80-cf3e-11eb-96c2-fa0e3a744837.png)


For summary:
- please add API section for react integration
- if needed, link to the newest API section in react docs
- could you explain to me sth more about the setup for react samples? Which node version you are using? I still wasn't able to run samples from react repo itself. I mean, the `build` and `start` commands look fine in the console, but browser still get me: 
![image](https://user-images.githubusercontent.com/6498567/122338134-91be7700-cf3f-11eb-9752-93dc5b95be3f.png)
but since it's not the main topic here, we can leave it for now. But I'm not sure if it's related to broken samples in docs on my PC? Should I have sth more preinstalled?

MMMalik commented:
Hi @sculpt0r,

as discussed, a follow-up task will be created to move documentation under API section.

The following fixes have been applied:
* formatting is fixed for mentioned code samples
* broken link is removed
* `rc` version of `ckeditor4-react` is added to dependencies

Please test now. With `rc` version already published you don't need to link local packages anymore.

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+
+### API
+
+The component's API exposes initial configuration options such as `config`, `editorUrl`, or `initData` as well as editor's event handlers. If you find `CKEditor` component's capabilities insufficient, consider using more flexible `useCKEditor` hook.
```
```suggestion
The component's API exposes initial configuration options such as `config`, `editorUrl` or `initData` as well as editor's event handlers. If you find `CKEditor` component's capabilities insufficient, consider using a more flexible `useCKEditor` hook.
```

MMMalik replied:
Similarly to `and` rule, putting comma before `or` is [correct](https://www.grammarly.com/blog/comma-before-or/). I will use it consistently.

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+| debug         | bool                  | `undefined`                              | Toggles debugging mode. Use it to log info related to editor events.                                                                                                                                                                                                                                                       ...
+| dispatchEvent | func                  | `undefined`                              | Dispatches editor events.<br> Pass a custom function or `dispatch` from `useReducer`. Event type and payload will be passed as part of the first argument: `({ type, payload }) => { ... }`. Event name is provided as `type` and its value is prefixed with `__CKE__` in order to facilitate integration with `useReducer ...
+| editorUrl     | string                | link to [CDN](https://cdn.ckeditor.com/) | See [Customizing CKEditor Preset or Version](#customizing-ckeditor-preset-or-version).                                                                                                                                                                                                                                     ...
+| element       | object                | `undefined`                              | Required prop. Ref to underlying to DOM element.                                                                                                                                                                                                                                                                           ...
```
Let's try to avoid abbreviations.
```suggestion
| element       | object                | `undefined`                              | Required. Reference to underlying to DOM element.                                                                                                                                                                                                                                                                                                                                              |
```

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
````diff
@@ -0,0 +1,526 @@
+export default Feedback;
+```
+
+For more examples see [repo](https://github.com/ckeditor/ckeditor4-react/tree/master/samples).
````
After the release of v2, the newest samples will be on `stable` branch. We should redirect users to the stable version of samples.
```suggestion
For more examples see [repo](https://github.com/ckeditor/ckeditor4-react/tree/stable/samples).
```

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+
+## Editor initialization
+
+Due to the nature of React functional components (including [hooks rules](https://reactjs.org/docs/hooks-rules.html)) some props are memoized under the hood. React integration ensures referential stability of the following props: `config`, `debug`, `type`, `editorUrl`, `dispatchEvent`, `subscribeTo`, and all event handlers (`on${EventName}`). Passing new values of these props across renders will  ...
```
```suggestion
Due to the nature of React functional components (including [hooks rules](https://reactjs.org/docs/hooks-rules.html)) some props are memoized under the hood. React integration ensures referential stability of the following props: `config`, `debug`, `type`, `editorUrl`, `dispatchEvent`, `subscribeTo` and all event handlers (`on${EventName}`). Passing new values of these props across renders will have no impact on already created instances of `CKEditor` component or `useCKEditor` hook. A new component instance must be created in order to override memoized values. This can be achieved by leveraging `key` prop.
```

MMMalik replied:
AFAIK, putting a comma before `and` in a list is [optional but correct](https://www.grammarly.com/blog/comma-before-and/).

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+
+### CKEditorEventAction
+
+Object that maps editor event names to their prefixed equivalents, e.g. `instanceReady` -> `__CKE__instanceReady`. It's useful when using `useCKEditor` in combination with `useReducer`.
```
I'm not fully convinced to use such `arrows` in the documentation.

```suggestion
Object that maps editor event names to their prefixed equivalents, e.g. `instanceReady` becomes `__CKE__instanceReady`. It's useful when using `useCKEditor` in combination with `useReducer`.
```

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+
+	const { editor } = useCKEditor( {
+		element,
+		// `dispatchEvent` is memoized, so initial value of `someProp` will be used
```
```suggestion
		// `dispatchEvent` is memoized, so initial value of `someProp` will be used.
```

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+	useCKEditor<'customEvent' | 'anotherCustomEvent'>( {
+		element,
+		dispatchEvent: ( { type } ) => {
+			if ( type === prefixEventName( 'customEvent' ) ) {
```
I'm thinking about:
- `prefixEventName( 'customEvent' )` should return `__CKE__customEvent`.
- So if `type` is equal to the `prefixEventName( 'customEvent' )`,  `console.log( type );` should return `__CKE__customEvent`, not `customEvent` ?

If I'm not missing something here, please correct strings placed as comments after both `console.log`-s.

MMMalik replied:
Indeed, that's my mistake.

sculpt0r commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+
+Other notable differences are:
+
+-   `editorUrl` is now passed as a prop
```
```suggestion
-   `editorUrl` is now passed as a prop,
```

sculpt0r reviewed (changes requested):
@MMMalik 

Thank you for all the previous corrects.

Now, I was able to properly build docs with samples - thank you for your efforts to release `RC`, so that makes things working.

I also test easy image which is provided in samples config, but I'm getting: 
![image](https://user-images.githubusercontent.com/6498567/122550125-9a907500-d033-11eb-9dc9-781f31a6a86d.png)
So maybe we could just remove it from there?

I left a few inline comments, but they are mostly some style corrects. So after that, I think it will be finally ready to merge ;) :+1: 

Also, I extracted an issue for API docs: https://github.com/ckeditor/ckeditor4-react/issues/225

MMMalik commented on docs/sdk/examples/react.html:
```diff
@@ -277,104 +279,135 @@ <h3>Events Log</h3>
+	 */
+	const setEditorData = useMemo( () => {
+		if ( editor ) {
+			/* eslint-disable-next-line */
```
Agreed, that's a copy-paste artifact ;-)

MMMalik commented on docs/guide/dev/integration/react/v2/README.md:
```diff
@@ -0,0 +1,526 @@
+
+| Prop             | Type                  | Default                                  | Description                                                                                                                                                                                                                                                                                                             ...
+| ---------------- | --------------------- | ---------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| config           | object                | `{}`                                     | Config object passed to editor's constructor.<br> See config details [here](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html).                                                                                                                                                                       ...
```
Fair point, thanks.

MMMalik commented on scripts/integrations/react/ConfigEvents.jsx:
```diff
@@ -1,79 +1,65 @@
 		} ).format( new Date() );
 
-		console.log( event.timestamp, event.name, event.data || 'No additional data was provided' );
+		console.log( evt.timestamp, evt.name, evt.data || 'No additional data was provided' );
```
Agreed, let's remove it.

MMMalik commented on docs/sdk/examples/react.html:
```diff
@@ -137,10 +137,10 @@
 			</style>
 
-			<h1>CKEditor 4 WYSIWYG Editor React Integration <a class="documentation" href="https://ckeditor.com/docs/ckeditor4/latest/guide/dev_react.html">Documentation</a></h1>
+			<h1>CKEditor 4 WYSIWYG Editor React Integration <a class="documentation" href="https://ckeditor.com/docs/ckeditor4/latest/guide/dev_react_v2.html">Documentation</a></h1>
```
Agreed!

MMMalik commented:
Hi @sculpt0r,

I have adjusted formatting and links as per your suggestions. 

Regarding EasyImage, I'm seeing that error in other integrations as well but only locally. It seems to be working fine in production though. My suggestion is to hold on with any changes related to that issue.

sculpt0r reviewed (approved):
LGTM 🎉 👍 

I think that all of your efforts finally leads us to the merge :) Thank you for all corrects.
